File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/NexusHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.handler.str
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import com.zeroday.zdscript.NexusScriptEngine
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.jsonObject

/**
 * Routes messages to the in-game Nexus scripting engine. Scripts can be
 * run inline, saved, listed, deleted, or syntax-validated without execution.
 */
class NexusHandler(
    private val scriptEngine: NexusScriptEngine
) : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.NEXUS_RUN,
        RequestTypes.NEXUS_SAVE,
        RequestTypes.NEXUS_LIST,
        RequestTypes.NEXUS_DELETE,
        RequestTypes.NEXUS_VALIDATE
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.NEXUS_RUN -> run(ctx, playerId, payload)
            RequestTypes.NEXUS_SAVE -> save(ctx, playerId, payload)
            RequestTypes.NEXUS_LIST -> listScripts(ctx, playerId)
            RequestTypes.NEXUS_DELETE -> delete(ctx, playerId, payload)
            RequestTypes.NEXUS_VALIDATE -> validate(payload)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported nexus request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun run(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val code = payload.str("code")
        val scriptId = payload.str("script_id")
        if (code == null && scriptId == null) {
            return HandlerResult.Err(ServerError.fromMessage("Provide 'code' or 'script_id'", ErrorCategory.VALIDATION))
        }
        val result = if (code != null) {
            ctx.services.hackToLearnService.runCode(playerId, code)
        } else {
            val argsJson = payload["args"]?.jsonObject
            val args: Map<String, String> = argsJson?.entries?.associate { (k, v) -> k to (v.jsonPrimitive.content) } ?: emptyMap()
            ctx.services.hackToLearnService.executeScript(playerId, scriptId!!, args)
        }
        return result.fold(
            onSuccess = { scriptResult ->
                HandlerResult.Ok(
                    type = ResponseTypes.NEXUS_RESULT,
                    payload = mapOf(
                        "output" to scriptResult.output,
                        "console" to scriptResult.console,
                        "variables" to scriptResult.variables,
                        "success" to true
                    )
                )
            },
            onFailure = { e ->
                HandlerResult.Err(
                    ServerError.fromMessage(e.message ?: "Execution failed", ErrorCategory.INTERNAL),
                    type = ResponseTypes.NEXUS_ERROR
                )
            }
        )
    }

    private suspend fun save(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val name = payload.str("name")
            ?: return HandlerResult.Err(ServerError.fromMessage("Script name required", ErrorCategory.VALIDATION))
        val code = payload.str("code")
            ?: return HandlerResult.Err(ServerError.fromMessage("Script code required", ErrorCategory.VALIDATION))
        val desc = payload.str("description") ?: ""
        return ctx.services.hackToLearnService.saveScript(playerId, name, code, desc).fold(
            onSuccess = { script -> HandlerResult.Ok(ResponseTypes.SCRIPT_SAVED, mapOf("script" to script)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Save failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun listScripts(ctx: HandlerContext, playerId: String): HandlerResult {
        val scripts = ctx.services.hackToLearnService.getPlayerScripts(playerId)
        return HandlerResult.Ok(ResponseTypes.SCRIPT_LIST, mapOf("scripts" to scripts))
    }

    private suspend fun delete(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val scriptId = payload.str("script_id")
            ?: return HandlerResult.Err(ServerError.fromMessage("Script ID required", ErrorCategory.VALIDATION))
        return ctx.services.hackToLearnService.deleteScript(playerId, scriptId).fold(
            onSuccess = { HandlerResult.Ok(ResponseTypes.SCRIPT_DELETED, mapOf("message" to "Script deleted")) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Delete failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun validate(payload: JsonObject): HandlerResult {
        val code = payload.str("code")
            ?: return HandlerResult.Err(ServerError.fromMessage("Code required", ErrorCategory.VALIDATION))
        val result = scriptEngine.validate(code)
        return HandlerResult.Ok(
            type = ResponseTypes.VALIDATION_RESULT,
            payload = mapOf(
                "valid" to result.valid,
                "errors" to result.errors
            )
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/NexusHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class NexusHandler : IHandler
{
    public string MessageType => "nexus";

    public NexusHandler()
    {
    }

    public Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Missing payload" }));

        var action = payload.Value.Str("action");
        return action switch
        {
            "run" => HandleRun(connectionId, payload.Value),
            "save" => HandleSave(connectionId, payload.Value),
            "list" => HandleList(connectionId, payload.Value),
            "delete" => HandleDelete(connectionId, payload.Value),
            "validate" => HandleValidate(connectionId, payload.Value),
            _ => Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = $"Unknown nexus action: {action}" }))
        };
    }

    private static Task<IActionResult> HandleRun(string connectionId, JsonElement payload)
    {
        var code = payload.Str("code");
        var scriptId = payload.Str("script_id");

        if (string.IsNullOrEmpty(code) && string.IsNullOrEmpty(scriptId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Provide 'code' or 'script_id'" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("nexus_result", connectionId, new
            {
                output = "Script executed (placeholder)",
                console = "",
                variables = new Dictionary<string, object>(),
                success = true
            }));
    }

    private static Task<IActionResult> HandleSave(string connectionId, JsonElement payload)
    {
        var name = payload.Str("name");
        var code = payload.Str("code");

        if (string.IsNullOrEmpty(name))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Script name required" }));
        if (string.IsNullOrEmpty(code))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Script code required" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("script_saved", connectionId, new
            {
                script = new
                {
                    id = Guid.NewGuid().ToString(),
                    name,
                    description = payload.Str("description") ?? ""
                }
            }));
    }

    private static Task<IActionResult> HandleList(string connectionId, JsonElement payload)
    {
        return Task.FromResult<IActionResult>(
            new MessageResult("script_list", connectionId, new
            {
                scripts = Array.Empty<object>()
            }));
    }

    private static Task<IActionResult> HandleDelete(string connectionId, JsonElement payload)
    {
        var scriptId = payload.Str("script_id");
        if (string.IsNullOrEmpty(scriptId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Script ID required" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("script_deleted", connectionId, new { message = "Script deleted" }));
    }

    private static Task<IActionResult> HandleValidate(string connectionId, JsonElement payload)
    {
        var code = payload.Str("code");
        if (string.IsNullOrEmpty(code))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Code required" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("validation_result", connectionId, new
            {
                valid = true,
                errors = Array.Empty<string>()
            }));
    }
}
